### ISSUE TITLE
To check: are tasks logs written at the end of tasks or during execution?

### ISSUE THREAD

**Issue by tcompa** [fractal-analytics-platform/fractal-server#888]
Relevant piece of code (from ` _call_command_wrapper`, in `runnner/_common.py`):
```python
    fp_stdout = open(stdout, "w")
    fp_stderr = open(stderr, "w")
    try:
        result = subprocess.run(  # nosec
            shlex_split(cmd),
            stderr=fp_stderr,
            stdout=fp_stdout,
        )
    except Exception as e:
        raise e
    finally:
        fp_stdout.close()
        fp_stderr.close()
```

Docs for `open` (https://docs.python.org/3/library/functions.html#open) say:

> buffering is an optional integer used to set the buffering policy. Pass 0 to switch buffering off (only allowed in binary mode), 1 to select line buffering (only usable in text mode), and an integer > 1 to indicate the size in bytes of a fixed-size chunk buffer. Note that specifying a buffer size this way applies for binary buffered I/O, but TextIOWrapper (i.e., files opened with mode='r+') would have another buffering. To disable buffering in TextIOWrapper, consider using the write_through flag for [io.TextIOWrapper.reconfigure()](https://docs.python.org/3/library/io.html#io.TextIOWrapper.reconfigure). When no buffering argument is given, the default buffering policy works as follows:
>
> * Binary files are buffered in fixed-size chunks; the size of the buffer is chosen using a heuristic trying to determine the underlying device’s “block size” and falling back on [io.DEFAULT_BUFFER_SIZE](https://docs.python.org/3/library/io.html#io.DEFAULT_BUFFER_SIZE). On many systems, the buffer will typically be 4096 or 8192 bytes long.
>
> * “Interactive” text files (files for which [isatty()](https://docs.python.org/3/library/io.html#io.IOBase.isatty) returns True) use line buffering. Other text files use the policy described above for binary files.

Thus I expect buffering to be active by default, with something like 4KB or 8KB buffer size. Some log files are smaller than that, and then they likely will only be written at the end. Larger log files (e.g. for cellpose) should be written in chunks during execution:

```
1.4K 0.err
   0 0.out
 989 1_par_20200812-CardiomyocyteDifferentiation14-Cycle1_zarr_B_03_0_.err
   0 1_par_20200812-CardiomyocyteDifferentiation14-Cycle1_zarr_B_03_0_.out
1.9K 2.err
   0 2.out
 525 3_par_20200812-CardiomyocyteDifferentiation14-Cycle1_mip_zarr_B_03_0_.err
   0 3_par_20200812-CardiomyocyteDifferentiation14-Cycle1_mip_zarr_B_03_0_.out
 11K 4_par_20200812-CardiomyocyteDifferentiation14-Cycle1_mip_zarr_B_03_0_.err
   0 4_par_20200812-CardiomyocyteDifferentiation14-Cycle1_mip_zarr_B_03_0_.out
3.5K 5_par_20200812-CardiomyocyteDifferentiation14-Cycle1_mip_zarr_B_03_0_.err
   0 5_par_20200812-CardiomyocyteDifferentiation14-Cycle1_mip_zarr_B_03_0_.out
```

**Comment by tcompa** [comment]
TLDR
We are probably buffering log files at 4K/8K buffer sizes, which means e.g. 100 log lines of 81 characters.
We can probably easily switch to single-line buffering (with `buffering=1`), but I don't know how to make it e.g. 1K (see example below).

---

```python
import io
import matplotlib.pyplot as plt

single_log_line = "Log line: this is a typical log line, with some contents and some more contents.\n"
lines = list(range(500))

for buffering in [-1, 1, 10]:
    print(f"{buffering=}")
    path = f"/tmp/xxx_{buffering=}"

    fp = open(path, "w", buffering=buffering)

    file_size = []
    for i in lines:
        fp.write(single_log_line)
        with open(path, "r") as fp2:
            file_size.append(len(fp2.read()))
    fp.close()
    plt.plot(lines, file_size, label=f"{buffering=}", alpha=0.8, lw=2)

plt.title(f"{io.DEFAULT_BUFFER_SIZE=}, {len(single_log_line)=}")
plt.yticks(range(0, 40000, 8192))
plt.xlabel("Number of log lines emitted")
plt.ylabel("File size in bytes")
plt.legend()
plt.savefig("fig.png", dpi=256)
```

![image](https://github.com/fractal-analytics-platform/fractal-server/assets/3862206/121dad79-077b-42df-bd36-ab0abddc541f)

**Comment by tcompa** [comment]
I'm closing this, as the buffering mechanism seems quite clear, and possible updates are defined.
Let's re-open if we think that we need to update our current 4K/8K buffering for logs.
